pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,618 @@
+type Manager interface {
+	PreFilter(context.Context, *corev1.Pod) error
+	Permit(context.Context, *corev1.Pod) (time.Duration, Status)
+	PostBind(context.Context, *corev1.Pod, string)
+	PostFilter(context.Context, *corev1.Pod, framework.Handle, string) (*framework.PostFilterResult, *framework.Status)
+	GetPodGroup(*corev1.Pod) (string, *v1alpha1.PodGroup)
+	GetCreatTime(*framework.QueuedPodInfo) time.Time
+	ActivateSiblings(*corev1.Pod, *framework.CycleState)
+	AllowGangGroup(*corev1.Pod, framework.Handle, string)
+	Unreserve(context.Context, *framework.CycleState, *corev1.Pod, string, framework.Handle, string)
+	Recover()
+}
+
+// PodGroupManager defines the scheduling operation called
+type PodGroupManager struct {
+	// pgClient is a podGroup client
+	pgClient pgclientset.Interface
+	// snapshotSharedLister is pod shared list
+	snapshotSharedLister framework.SharedLister
+	// scheduleTimeout is the default timeout for podgroup scheduling.
+	// If podgroup's scheduleTimeoutSeconds is set, it will be used.
+	scheduleTimeout *time.Duration
+	// pgLister is podgroup lister
+	pgLister pglister.PodGroupLister
+	// podLister is pod lister
+	podLister listerv1.PodLister
+	// reserveResourcePercentage is the reserved resource for the max finished group, range (0,100]
+	reserveResourcePercentage int32
+	// cache stores gang info
+	cache *GangCache
+	sync.RWMutex
+}
+
+// NewPodGroupManager creates a new operation object.
+func NewPodGroupManager(pgClient pgclientset.Interface, snapshotSharedLister framework.SharedLister,
+	pgInformer pginformer.PodGroupInformer, podInformer informerv1.PodInformer, args *config.CoschedulingArgs) *PodGroupManager {
+	pgMgr := &PodGroupManager{
+		pgClient:             pgClient,
+		snapshotSharedLister: snapshotSharedLister,
+		pgLister:             pgInformer.Lister(),
+		podLister:            podInformer.Lister(),
+	}
+	gangCache := NewGangCache(args, podInformer.Lister(), pgInformer.Lister())
+	pgMgr.cache = gangCache
+	return pgMgr
+}
+
+func (pgMgr *PodGroupManager) OnPodAdd(obj interface{}) {
+	pgMgr.cache.onPodAdd(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodDelete(obj interface{}) {
+	pgMgr.cache.onPodDelete(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodGroupAdd(obj interface{}) {
+	pgMgr.cache.onPodGroupAdd(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodGroupDelete(obj interface{}) {
+	pgMgr.cache.onPodGroupDelete(obj)
+}
+
+func (pgMgr *PodGroupManager) Recover() {
+	pgMgr.cache.RecoverGangCache()
+}
+
+// ActivateSiblings stashes the pods belonging to the same PodGroup of the given pod
+// in the given state, with a reserved key "kubernetes.io/pods-to-activate".
+func (pgMgr *PodGroupManager) ActivateSiblings(pod *corev1.Pod, state *framework.CycleState) {
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	pods, err := pgMgr.podLister.Pods(pod.Namespace).List(labels.NewSelector())
```

Comment:
you can get all pods from gangCache's gang

Reply:
yes, active all the gang from the gangGroup will make the pods waiting on Permit stage more efficient